style/javascript.md

```diff
@@ -536,6 +537,404 @@ to replace all uses of `var` in existing files.
+
+There are a couple methods that are sufficiently complicated and don't have a direct equivalent so instead we have a custom-built copy of lodash containing only those methods. You can find this file at: `third_party/javascript-khansrc/lodash/lodash.js` along with instructions on how to build it and exactly what methods are included.
+
+What follows is a method-by-method set of equivalents for what we've used in Underscore and what you should be using in ES6/7 instead:
+
+#### _.bind
+
+```js
+_.bind(fn, someObj, args)
+```
+
+Becomes:
+```js
+fn.bind(someObj, args)
+```
+
+Note that cases where you had:
+
+```js
+_.bind(function() { ... }, this)
+```
+
+should instead become:
+
+```js
+() => { ... }
+```
+
+#### _.bindAll
+
+```js
+_.bindAll(someObj, "methodName")
+```
+
+Becomes:
+```js
+someObj.methodName = someObj.methodName.bind(someObj);
+```
+
+(or use a loop if binding multiple methods)
+
+#### _.clone
+
+No alternative at the moment! If you need it then you should add it to the compiled version of lodash and then update this guide to mention that it now exists!
+
+#### _.debounce
+
+This is included in our special-built version of lodash.
+
+#### _.defer
+
+```js
+_.defer(function() {
+    foo();
+});
+```
+
+Becomes:
+```js
+setTimeout(function() {
+    foo();
+}, 0);
+```
+
+#### _.delay
+
+```js
+_.delay(function() {
+    loadJavaScript();
+    requireSuggestions();
+}, 2000);
+```
+
+Becomes:
+```js
+setTimeout(function() {
+    loadJavaScript();
+    requireSuggestions();
+}, 2000);
+```
+
+#### _.each
+
+```js
+_.each(array)
+```
+
+Becomes:
+```js
+for (let val of array) {
+
+}
+```
+
+or:
+```js
+array.forEach(...)
+```
+
+```js
+_.each(object)
+```
+
+Becomes:
+```js
+for (let [key, val] of Object.entries(object)) {
+
+}
+```
+
+#### _.extend
+
+```js
+_.extend({}, dataToEventLogOnly, dataToGraphite,
+        {_request_id: KA.requestLogId,
+         _graphite_key_prefix: KA.gaeStatsKeyPrefix,
+         _graphite_keys: Object.keys(dataToGraphite).join()
+        })
+```
+
+Becomes:
+```js
+{
+    ...dataToEventLogOnly,
+    ...dataToGraphite,
+    _request_id: KA.requestLogId,
+    _graphite_key_prefix: KA.gaeStatsKeyPrefix,
+    _graphite_keys: Object.keys(dataToGraphite).join()
+}
+```
+
+or:
+```js
+_.extend(defaultOptions, options || {})
+```
+
+Becomes:
+```js
+{...defaultOptions, ...options}
+```
+
+or:
+```js
+_.extend(json, this.model.toJSON())
+```
+
+Becomes:
+```js
+Object.assign(json, this.model.toJSON())
+```
+
+#### _.filter
+
+```js
+_.filter(resourceTimings, function(t) {
+    return nameRegexp.test(t.name);
+})
+```
+
+Becomes:
+```js
+resourceTimings.filter(t => nameRegexp.test(t.name))
+```
+
+#### _.has
+
+```js
+_.has(array, value)
+```
+
+Becomes:
+```js
+array.includes(value)
+```
+
+or:
+```js
+_.has(obj, value)
+```
+
+Becomes:
+```js
+value in obj
```

Comment:
@ariabuckles I agree that this could be a problem in some very rare cases however I think this is more of a concern for libraries that need to live in every conceivable environment (like jQuery, for example, where stuff like this was a big concern -- maybe even something like tota11y or KaTeX). But for development in the khan-webapp repo I suspect that that's much less of a concern (not to mention that simple presence checks should ideally be as fast as possible - piping them through a method like `_.has()` tends to negate that). I'll add a note emphasizing this point so that people don't blindly use it.